### PR TITLE
Rough draft of IOC data model.

### PR DESCRIPTION
**Pull request by AWSHurneyt**
### Description
These are the references I'm using.

**Domain documentation**
Section "6.4 Domain Name Object" of https://docs.oasis-open.org/cti/stix/v2.1/os/stix-v2.1-os.html#_k017w16zutw

**Hash documentation**
https://stix2.readthedocs.io/en/latest/guide/serializing.html

**IP documentation**
"6.8 IPv4 Address Object" section of https://docs.oasis-open.org/cti/stix/v2.1/os/stix-v2.1-os.html#_wh296fiwpklp
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by eirsep — CHANGES_REQUESTED** [review]
can we add a validate function to verify feed id, values, type and value are non-null

**Comment by AWSHurneyt** [comment]
> can we add a validate function to verify feed id, values, type and value are non-null

@eirsep Added some basic validation.

File: src/main/java/org/opensearch/securityanalytics/model/IocDao.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.core.xcontent.XContentParserUtils;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+import static org.opensearch.securityanalytics.model.Detector.NO_ID;
```
Comment (jowg-amazon) [inline src/main/java/org/opensearch/securityanalytics/model/IocDao.java:24]: This references detector's No ID field, might be better to have IocDao's own field.

File: src/main/java/org/opensearch/securityanalytics/model/IocDto.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                .field(IocDao.SEVERITY_FIELD, severity)
+                .field(IocDao.SPEC_VERSION_FIELD, specVersion)
+                .timeField(IocDao.CREATED_FIELD, created)
+                .timeField(IocDao.MODIFIED_FIELD, modified)
+                .field(IocDao.DESCRIPTION_FIELD, description)
+                .field(IocDao.LABELS_FIELD, labels)
+                .field(IocDao.FEED_ID_FIELD, feedId)
+                .endObject();
+    }
+}
```
Comment (jowg-amazon) [inline src/main/java/org/opensearch/securityanalytics/model/IocDto.java:83]: Does the DTO need a parse function?

File: src/main/java/org/opensearch/securityanalytics/model/IocDao.java (hunk 1 above)
Comment (AWSHurneyt) [reply src/main/java/org/opensearch/securityanalytics/model/IocDao.java]: Added a `NO_ID` variable in `IocDao`.

File: src/main/java/org/opensearch/securityanalytics/model/IocDto.java (hunk 2 above)
Comment (AWSHurneyt) [reply src/main/java/org/opensearch/securityanalytics/model/IocDto.java:140]: Added a `parse` function.

**Review by eirsep — COMMENTED** [review]
LGTM 
plz add serialization tests

**Comment by AWSHurneyt** [comment]
> LGTM plz add serialization tests



> LGTM plz add serialization tests

@eirsep Added some unit tests. Will add more once we're more confident that this is the final structure of the data model.

**Review by eirsep — APPROVED** [review]

**Pull request merged**